tpp-app/src/services/CalculationService.js

```diff
@@ -0,0 +1,162 @@
+    const twoDaysEarlierSymptoms = lastThreeDaysSymptoms[0];
+    const yesterdaySymptoms = lastThreeDaysSymptoms[1];
+    const todaySymptoms = lastThreeDaysSymptoms[2];
+
+    const todayNoPeriod = (todaySymptoms.flow == null || todaySymptoms.flow === FLOW_LEVEL.NONE);
+    const yesterdayNoPeriod = (yesterdaySymptoms.flow == null || yesterdaySymptoms.flow === FLOW_LEVEL.NONE);
+    const twoDaysEarlierPeriod = (twoDaysEarlierSymptoms.flow != null && twoDaysEarlierSymptoms.flow !== FLOW_LEVEL.NONE);
+    return todayNoPeriod && yesterdayNoPeriod && twoDaysEarlierPeriod;
+};
+
+/**
+ * Returns True if today is the start of a period, False otherwise i.e. matches period pattern of _ _ X
+ * @return {boolean} Returns a boolean indicating if period has started today
+ */
+export const isPeriodStarting = async () => {
+    const lastThreeDaysSymptoms = await GETLastThreeDaysSymptoms();
+    const twoDaysEarlierSymptoms = lastThreeDaysSymptoms[0];
+    const yesterdaySymptoms = lastThreeDaysSymptoms[1];
+    const todaySymptoms = lastThreeDaysSymptoms[2];
+
+    const todayPeriod = (todaySymptoms.flow != null && todaySymptoms.flow !== FLOW_LEVEL.NONE);
+    const yesterdayNoPeriod = (yesterdaySymptoms.flow == null || yesterdaySymptoms.flow === FLOW_LEVEL.NONE);
+    const twoDaysEarlierNoPeriod = (twoDaysEarlierSymptoms.flow == null || twoDaysEarlierSymptoms.flow === FLOW_LEVEL.NONE);
+    return todayPeriod && yesterdayNoPeriod && twoDaysEarlierNoPeriod;
+}
+
+/**
+ * Calculates the average period length of the user.
+ */
+export const calculateAveragePeriodLength = async () => new Promise( async (resolve, reject) => {
+    // Only calculate Average Period Length once period is over
+    if (isPeriodOver()) {
+        // Record all period intervals
+        let completeHistory = [];
+
+        // Get all Years
+        GETStoredYears()
+            .then((years) => {
+                years.forEach(async (year) => {
+                    completeHistory.push.apply(completeHistory, await CycleService.GETCycleHistoryByYear(year));
+                });
+            })
+            .catch((e) => {
+                console.log(`calculateAveragePeriodLength error: ${JSON.stringify(error)}`);
+                reject();
+            });
+
+        // Use reduce to find average
+        const averagePeriodLength = completeHistory.reduce(function (sum, interval) {
+            return sum + interval.periodDays;
+        }, 0) / completeHistory.length;
+
+        await AsyncStorage.Set(
+            KEYS.AVERAGE_PERIOD_LENGTH, JSON.stringify(averagePeriodLength)
+        ).then(() => {
+            console.log(`Recalculated ${KEYS.AVERAGE_PERIOD_LENGTH} as ${averagePeriodLength}`)
+            resolve();
+        }).catch((error) => {
+            console.log(`calculateAveragePeriodLength error: ${JSON.stringify(error)}`);
+            reject();
+        });
+    }
+
+    // Else don't calculate, do nothing
```

Comment:
Should conclude the Promise here just to make sure all cases in the function end with either a `reject` or `resolve`.